platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/service/WritableRosterStore.java

```diff
@@ -0,0 +1,225 @@
+                    lastUsedActiveRoster, "Active Roster must be present in the state during normal network restart.");
+        }
+
+        // software upgrade is detected, we adopt the candidate roster
+        adoptCandidateRoster(initialState.get().getRound() + 1);
+        commit();
+        return candidateRoster;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Roster getCandidateRoster() {
+        return rosterStateAccessor.getCandidateRoster();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Roster getActiveRoster() {
+        return rosterStateAccessor.getActiveRoster();
+    }
+
+    /**
+     * Adopts the candidate roster present in the state as the active roster.
+     * Until the Dynamic Address Book is implemented, this method will be called unconditionally on network upgrade.
+     *
+     * @param roundNumber the round number in which the candidate roster became active
+     */
+    private void adoptCandidateRoster(final long roundNumber) {
+        final RosterState previousRosterState = rosterStateOrThrow();
+        final Roster candidateRoster = rosterMap.get(ProtoBytes.newBuilder()
+                .value(previousRosterState.candidateRosterHash())
+                .build());
+        if (candidateRoster == null) {
+            throw new IllegalStateException("Candidate roster not found in the state.");
+        }
+        RosterValidator.validate(candidateRoster);
+
+        storeAsActive(candidateRoster, roundNumber);
+        removeCandidateRoster();
+    }
+
+    /**
+     * Stores this roster as the active roster.
+     *
+     * @param roster        a roster to set as active
+     * @param round        the round in which this roster became active
+     */
+    private void storeAsActive(@NonNull final Roster roster, final long round) {
+        Objects.requireNonNull(roster);
+        RosterValidator.validate(roster);
+
+        final RosterState previousRosterState = rosterStateOrThrow();
+
+        // update the roster state
+        final List<RoundRosterPair> roundRosterPairs = new LinkedList<>(previousRosterState.roundRosterPairs());
+        final Bytes activeRosterHash = RosterUtils.hashOf(roster).getBytes();
+        roundRosterPairs.addFirst(new RoundRosterPair(round, activeRosterHash));
+
+        // remove the formerly previous active roster, i.e., the roster that was active before the last two adopted
+        // rosters, if any.
+        if (roundRosterPairs.size() > 2) {
+            roundRosterPairs.removeLast();
+        }
+
+        final Builder rosterStateBuilder = RosterState.newBuilder()
+                .candidateRosterHash(previousRosterState.candidateRosterHash())
+                .roundRosterPairs(roundRosterPairs);
+        this.rosterState.put(rosterStateBuilder.build());
+
+        // update the roster map
+        this.rosterMap.put(ProtoBytes.newBuilder().value(activeRosterHash).build(), roster);
```

Comment:
Since the active roster being adopted always comes from the candidate roster, shouldn't it already be in the roster map?

Reply:
done